File: ruby/tournament/_iterations/2/tournament.rb
module Tournament
  HEADER = "Team                           | MP |  W |  D |  L |  P\n"

  class Team
    attr_accessor :wins, :draws, :losses

    def initialize
      @wins = @draws = @losses = 0
    end

    def played
      wins + draws + losses
    end

    def points
      wins * 3 + draws
    end
  end

  def self.tally(input)
    results = Hash.new { |hash, name| hash[name] = Team.new }
    input.lines.each do |game_input|
      if /^(?<home>[\w ]+);(?<away>[\w ]+);(?<result>win|loss|draw)$/ =~ game_input
        home_team = results[home]
        away_team = results[away]
        case result
        when "win"
          home_team.wins += 1
          away_team.losses += 1
        when "loss"
          home_team.losses += 1
          away_team.wins += 1
        when "draw"
          home_team.draws += 1
          away_team.draws += 1
        end
      end
    end

    sorted_results = results.to_a.sort do |(n1, t1), (n2, t2)|
      cmp = t2.points <=> t1.points
      cmp == 0 ? n1 <=> n2 : cmp
    end
    HEADER + sorted_results.map do |(name, team)|
      "#{name.ljust(31, ' ')}| #{team.played.to_s.rjust(2, ' ')} | #{team.wins.to_s.rjust(2, ' ')} | #{team.draws.to_s.rjust(2, ' ')} | #{team.losses.to_s.rjust(2, ' ')} | #{team.points.to_s.rjust(2, ' ')}\n"
    end.join
  end
end

File: java/tournament/_iterations/1/src/main/java/Tournament.java
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Tournament {
    private static final Pattern RESULT_REGEX = Pattern.compile("^([^;]+);([^;]+);(win|loss|draw)$");

    private Map<String, TeamResults> results = new HashMap<>();

    void applyResults(String results) {
        String[] lines = results.split("\n");
        for (String line : lines) {
            Matcher matcher = RESULT_REGEX.matcher(line);
            if (matcher.find()) {
                String result = matcher.group(3);
                if (result.equals("win")) {
                    recordWin(matcher.group(1), matcher.group(2));
                } else if (result.equals("loss")) {
                    recordWin(matcher.group(2), matcher.group(1));
                } else {
                    recordDraw(matcher.group(1), matcher.group(2));
                }
            }
        }
    }

    String printTable() {
        List<String> lines = new ArrayList<>(
                Collections.singletonList("Team                           | MP |  W |  D |  L |  P"));
        this.results.entrySet().stream()
                               .map(entry -> new TeamWithResults(entry.getKey(), entry.getValue()))
                               .sorted((team1, team2) -> {
                                   int cmp = team2.results.points() - team1.results.points();
                                   if (cmp == 0) {
                                       cmp = team1.teamName.compareTo(team2.teamName);
                                   }
                                   return cmp;
                               }).forEach(team -> lines.add(String.format("%-31s|  %d |  %d |  %d |  %d |  %d",
                                                                            team.teamName, team.results.matchesPlayed(), team.results.wins(),
                                                                            team.results.draws(), team.results.losses(), team.results.points())));
        return String.join("\n", lines) + "\n";
    }

    private void recordWin(String winners, String losers) {
        this.results.computeIfAbsent(winners, n -> new TeamResults()).recordWin();
        this.results.computeIfAbsent(losers, n -> new TeamResults()).recordLoss();
    }

    private void recordDraw(String team1, String team2) {
        this.results.computeIfAbsent(team1, n -> new TeamResults()).recordDraw();
        this.results.computeIfAbsent(team2, n -> new TeamResults()).recordDraw();
    }

    private static final class TeamResults {
        private int wins;
        private int losses;
        private int draws;

        int matchesPlayed() {
            return this.wins + this.losses + this.draws;
        }

        int wins() {
            return this.wins;
        }

        int losses() {
            return this.losses;
        }

        int draws() {
            return this.draws;
        }

        int points() {
            return this.wins * 3 + this.draws;
        }

        void recordWin() {
            ++this.wins;
        }

        void recordLoss() {
            ++this.losses;
        }

        void recordDraw() {
            ++this.draws;
        }
    }

    private static final class TeamWithResults {
        final String teamName;
        final TeamResults results;

        TeamWithResults(String teamName, TeamResults results) {
            this.teamName = teamName;
            this.results = results;
        }
    }
}
